service/history/queue/split_policy.go

```diff
@@ -0,0 +1,276 @@
+// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+// THE SOFTWARE.
+
+package queue
+
+import (
+	"github.com/uber/cadence/service/history/task"
+)
+
+type (
+	lookAheadFunc func(task.Key) task.Key
+
+	pendingTaskSplitPolicy struct {
+		pendingTaskThreshold map[int]int // queue level -> threshold
+		lookAheadFunc        lookAheadFunc
+		maxNewQueueLevel     int
+	}
+
+	stuckTaskSplitPolicy struct {
+		attemptThreshold map[int]int // queue level -> threshold
+		maxNewQueueLevel int
+	}
+
+	selectedDomainSplitPolicy struct {
+		domainIDs     map[string]struct{}
+		newQueueLevel int
+	}
+
+	aggregatedSplitPolicy struct {
+		policies []ProcessingQueueSplitPolicy
+	}
+)
+
+// NewPendingTaskSplitPolicy creates a new processing queue split policy
+// based on the number of pending tasks
+func NewPendingTaskSplitPolicy(
+	pendingTaskThreshold map[int]int,
+	lookAheadFunc lookAheadFunc,
+	maxNewQueueLevel int,
+) ProcessingQueueSplitPolicy {
+	return &pendingTaskSplitPolicy{
+		pendingTaskThreshold: pendingTaskThreshold,
+		lookAheadFunc:        lookAheadFunc,
+		maxNewQueueLevel:     maxNewQueueLevel,
+	}
+}
+
+// NewStuckTaskSplitPolicy creates a new processing queue split policy
+// based on the number of task attempts tasks
+func NewStuckTaskSplitPolicy(
+	attemptThreshold map[int]int,
+	maxNewQueueLevel int,
+) ProcessingQueueSplitPolicy {
+	return &stuckTaskSplitPolicy{
+		attemptThreshold: attemptThreshold,
+		maxNewQueueLevel: maxNewQueueLevel,
+	}
+}
+
+// NewSelectedDomainSplitPolicy creates a new processing queue split policy
+// that splits out specific domainIDs
+func NewSelectedDomainSplitPolicy(
+	domainIDs map[string]struct{},
+	newQueueLevel int,
+) ProcessingQueueSplitPolicy {
+	return &selectedDomainSplitPolicy{
+		domainIDs:     domainIDs,
+		newQueueLevel: newQueueLevel,
+	}
+}
+
+// NewAggregatedSplitPolicy creates a new processing queue split policy
+// that which combines other policies. Policies are evaluated in the order
+// they passed in, and if one policy returns an non-empty result, that result
+// will be returned as is and policies after that one will not be evaluated
+func NewAggregatedSplitPolicy(
+	policies ...ProcessingQueueSplitPolicy,
+) ProcessingQueueSplitPolicy {
+	return &aggregatedSplitPolicy{
+		policies: policies,
+	}
+}
+
+func (p *pendingTaskSplitPolicy) Evaluate(
+	queue ProcessingQueue,
+) []ProcessingQueueState {
+	queueImpl := queue.(*processingQueueImpl)
+
+	threshold, ok := p.pendingTaskThreshold[queueImpl.state.level]
+	if !ok {
+		// no threshold specified for the level, skip spliting
+		return nil
+	}
+
+	if queueImpl.state.level == p.maxNewQueueLevel {
```

Comment:
Is the maxNewQueueLevel inclusive or exclusive?

Reply:
It's inclusive. For this split policy, if the current level is `x`, then the level for the split(new) queue will be `x+1`. So if the current level already equals to `maxNewQueueLevel` and you do split, level for new queue will exceed the threshold. Therefore, we need to skip splitting here.

Reply:
Actually, I can move this check before the check for `pendingTaskThreshold` to avoid confusion. But the result will be the same when current level = maxNewQueueLevel, either the split is rejected due to no threshold defined or even if there's one, it will be rejected by the maxNewQueueLevel.